### PR TITLE
Security

### PR DESCRIPTION
**Pull request by TarasOrlovskyi**

### CODE REVIEW

**Review by AlexKievDeveloper — DISMISSED** [review]

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java [hunk 1]
```diff
@@ -36,22 +34,26 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+                    "passed email and password to db {'email':{}, 'isAdded':{}}", email, isAdded);
+            if (isAdded) {
+                response.setStatus(HttpServletResponse.SC_SEE_OTHER);
+                logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_SEE_OTHER);
+                attributes.put("message", "Please confirm your registration. To do this, follow the link that we sent you by email.");
+                PageGenerator.getInstance().process("registration", attributes, response.getWriter());
+            } else {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
+                attributes.put("message", "Sorry, but a user with this email address already exists!");
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java]: "user already exists" shouldn't be displayed: it's a security issue.
Just "Sorry, couldn't sign you in. Check email and password?" or something.

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java [hunk 2]
```diff
@@ -36,22 +34,26 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-            response.setStatus(HttpServletResponse.SC_SEE_OTHER);
-            logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_SEE_OTHER);
-            response.sendRedirect(successRedirect);
-        } else {
+        if (!password.equals(confirmPassword)) {
             response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
             logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
-            //response.sendRedirect(failRedirect);
-            attributes.put("message", "Sorry, but user with this email already exist!");
+            attributes.put("message", "Sorry, the passwords do not match!");
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java]: This shouldn't be checked inside servlet, allowing submit only when passwords match is a task for front.
Here you'll find a couple answers without javascript: https://stackoverflow.com/questions/9142527/can-you-require-two-form-fields-to-match-with-html5

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java [hunk 3]
```diff
@@ -12,25 +12,39 @@
...
+    HttpServletRequest mockedRequest = mock(HttpServletRequest.class);
+    HttpServletResponse mockedResponse = mock(HttpServletResponse.class);
+    HttpSession mockedHttpSession = mock(HttpSession.class);
+    SignOutServlet signOutServlet = new SignOutServlet();
+    InOrder inOrderRequest = inOrder(mockedRequest);
+    InOrder inOrderResponse = inOrder(mockedResponse);
     @Test
-    @DisplayName("Checks if all right methods are called")
-    void doGetTest() throws IOException {
+    @DisplayName("Checks if all right methods are called & exist session was delete")
```
Comment (Greengrer) [inline src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java]: *deleteD.

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java [hunk 4]
```diff
@@ -12,25 +12,39 @@
...
+    void doGetWithExistSessionTest() throws IOException {
```
Comment (Greengrer) [inline src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java]: ExistingSession

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java (hunk 3 above)
Comment (Greengrer) [reply src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java]: existing session: існуюча сесія була видалена.

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java [hunk 5]
```diff
@@ -12,25 +12,39 @@
...
-        inOrderRequest.verify(mockedRequest).getSession();
+        inOrderRequest.verify(mockedRequest).getSession(false);
         verify(mockedHttpSession).invalidate();
         inOrderResponse.verify(mockedResponse).setStatus(HttpServletResponse.SC_OK);
         inOrderResponse.verify(mockedResponse).sendRedirect("/");
     }
+
+    @Test
+    @DisplayName("Checks if all right methods are called & session doesn't exist")
+    void doGetWithNonExistSessionTest() throws IOException {
```
Comment (Greengrer) [inline src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java]: NotExisting
or
NonExistent.

**Review by Greengrer — CHANGES_REQUESTED** [review]
I might be wrong about the "exist session" part.

File: src/main/resources/templates/registration.html [hunk 6]
```diff
@@ -1,163 +1,190 @@
...
+                                    alt="sign up">
+                        </a>
+                    </div>
+                    <div th:case="*">
+                        <a href="/signOut" class="head__login-buttons">
+                            <span class="tooltiptext">LogOut</span>
```
Comment (Greengrer) [inline src/main/resources/templates/registration.html:63]: *Log out is better than LogOut

File: src/main/resources/templates/registration.html [hunk 7]
```diff
@@ -1,163 +1,190 @@
...
+                                            document.getElementById('user_message').innerHTML="Sorry, the passwords don't match!";
+                                            document.form_reg.confirmPassword.focus();
+                                            return false;
+                                        }
+                                        document.form_reg.submit();
+                                    }
+                                </script>
+                                <label class="user__label">
+                                    <input class="user__input" type="text" name="discogsUserName"
+                                           placeholder="Discogs User Name">
```
Comment (Greengrer) [inline src/main/resources/templates/registration.html:166]: Add explanation what Discogs username is for, maybe?

Comment (Greengrer) [inline src/main/resources/templates/registration.html:166]: "When referring to an identity used with a password, "username" is used. If you are referring to a real name or full name, "user name" could be used. As far as "user-name," it is rarely used. In our earlier example, "John Smith" is the user name, and "smitj" is the username used with his password." (https://www.computerhope.com/jargon/u/username.htm)

Username, not User Name

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java [hunk 8]
```diff
@@ -16,43 +16,38 @@
...
             logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_SEE_OTHER);
-            response.sendRedirect(successRedirect);
+            attributes.put("message", "Please confirm your registration. To do this, follow the link that we sent you by email.");
+            PageGenerator.getInstance().process("registration", attributes, response.getWriter());
         } else {
             response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
             logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
-            //response.sendRedirect(failRedirect);
-            attributes.put("message", "Sorry, but user with this email already exist!");
+            attributes.put("message", "Sorry, but the email couldn't be registered. Check email and password!");
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java]: Я бы поменял email на user:
Sorry, but the user couldn't be registered.

File: src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java [hunk 9]
```diff
@@ -12,25 +12,39 @@
...
+        //prepare
+        when(mockedRequest.getSession(false)).thenReturn(null);
+        //when
+        signOutServlet.doGet(mockedRequest, mockedResponse);
+        //then
+        inOrderResponse.verify(mockedResponse).setContentType("text/html;charset=utf-8");
+        inOrderRequest.verify(mockedRequest).getSession(false);
+        verify(mockedHttpSession, times(0)).invalidate();
+        inOrderResponse.verify(mockedResponse).setStatus(HttpServletResponse.SC_OK);
+        inOrderResponse.verify(mockedResponse).sendRedirect("/");
```
Comment (Greengrer) [inline src/test/java/com/vinylteam/vinyl/web/servlets/SignOutServletTest.java:60]: Reusing same mock with inOrder is not secure.
(1)mockedResponse.setStatus(...);
(2)mockedResponse.sendRedirect(...);
(3)inOrderResponse.verify(...).setStatus(...);
(4)inOrderResponse.verify(...).sendRedirect(...);
(5)mockedResponse.setStatus(...);
(6)mockedResponse.sendRedirect(...);
(7)inOrderResponse.verify(...).setStatus(...);
(8)inOrderResponse.verify(...).sendRedirect(...);

The same as:

(1)mockedResponse.setStatus(...);
(2)mockedResponse.sendRedirect(...);
(5)mockedResponse.setStatus(...);
(6)mockedResponse.sendRedirect(...);
(3)inOrderResponse.verify(...).setStatus(...);
(4)inOrderResponse.verify(...).sendRedirect(...);
(7)inOrderResponse.verify(...).setStatus(...);
(8)inOrderResponse.verify(...).sendRedirect(...);

It has a potential to jumble tests up. Better to have @BeforeEach that nulls them and reassigns new mocks.

**Review by Greengrer — CHANGES_REQUESTED** [review]
That was cool!

**Review by AlexKievDeveloper — DISMISSED** [review]

**Review by Greengrer — APPROVED** [review]

**Review by AlexKievDeveloper — APPROVED** [review]

**Pull request merged**